### ISSUE TITLE
CWE-479 in CRT.c

### ISSUE THREAD

**Issue by BenBE** [htop-dev/htop#1563]
While compiling I noticed the following warning for [CWE-479](https://cwe.mitre.org/data/definitions/479.html) from GCC 14:

```
depbase=`echo CRT.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc-14 -DHAVE_CONFIG_H -I.  -DNDEBUG  -std=c99 -pedantic -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600  -I/usr/include/libnl3 -Wall -Wcast-align -Wcast-qual -Wextra -Wfloat-equal -Wformat=2 -Winit-self -Wmissing-format-attribute -Wmissing-noreturn -Wmissing-prototypes -Wpointer-arith -Wshadow -Wstrict-prototypes -Wundef -Wunused -Wwrite-strings -Wnull-dereference -D_XOPEN_SOURCE_EXTENDED -DSYSCONFDIR="\"/usr/local/etc\"" -I"./linux" -fanalyzer -MT CRT.o -MD -MP -MF $depbase.Tpo -c -o CRT.o CRT.c &&\
mv -f $depbase.Tpo $depbase.Po
CRT.c: In function ‘CRT_handleSIGTERM’:
CRT.c:846:4: warning: call to ‘snprintf’ from within signal handler [CWE-479] [-Wanalyzer-unsafe-call-within-signal-handler]
  846 |    snprintf(err_buf, sizeof(err_buf),
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  847 |            "A signal %d (%s) was received, exiting without persisting settings to htoprc.\n",
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  848 |            sgn, signal_str);
      |            ~~~~~~~~~~~~~~~~
  ‘CRT_installSignalHandlers’: events 1-2
    |
    |  952 | static void CRT_installSignalHandlers(void) {
    |      |             ^~~~~~~~~~~~~~~~~~~~~~~~~
    |      |             |
    |      |             (1) entry to ‘CRT_installSignalHandlers’
    |......
    |  966 |    signal(SIGINT, CRT_handleSIGTERM);
    |      |    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    |      |    |
    |      |    (2) registering ‘CRT_handleSIGTERM’ as signal handler
    |
  event 3
    |
    |cc1:
    | (3): later on, when the signal is delivered to the process
    |
    +--> ‘CRT_handleSIGTERM’: events 4-7
           |
           |  835 | static void CRT_handleSIGTERM(int sgn) {
           |      |             ^~~~~~~~~~~~~~~~~
           |      |             |
           |      |             (4) entry to ‘CRT_handleSIGTERM’
           |......
           |  838 |    if (!CRT_settings->changed)
           |      |       ~      
           |      |       |
           |      |       (5) following ‘false’ branch...
           |......
           |  841 |    const char* signal_str = strsignal(sgn);
           |      |                             ~~~~~~~~~~~~~~
           |      |                             |
           |      |                             (6) ...to here
           |......
           |  846 |    snprintf(err_buf, sizeof(err_buf),
           |      |    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           |      |    |
           |      |    (7) call to ‘snprintf’ from within signal handler
           |  847 |            "A signal %d (%s) was received, exiting without persisting settings to htoprc.\n",
           |      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           |  848 |            sgn, signal_str);
           |      |            ~~~~~~~~~~~~~~~~
           |
```

**Comment by cgzones** [comment]
Out of curiosity I asked an AI:

Based on the search results, snprintf is not considered async-signal-safe in C. Here’s a breakdown of the reasons:

* Valgrind bug: The snprintf function may not work correctly when called from a signal handler due to valgrind’s internal implementation. This is a specific issue with valgrind, not a fundamental problem with snprintf.
* ABI requirements: The x86-64 ABI requires floating-point arguments to be passed in XMM registers, which must be 16-byte aligned. Signal handlers may not provide this alignment, making snprintf unreliable.
* Non-async-signal-safe functions: snprintf is not listed as an async-signal-safe function in the POSIX standard or other reliable sources. This implies that it may not be safe to use in signal handlers.
* Initialization of locale machinery: snprintf may need to initialize locale machinery, which is not async-signal-safe.

To work around these issues, consider the following alternatives:

* Use a limited, async-signal-safe version of snprintf: The safe_snprintf function provided by idning/safe_snprintf is a simple, limited version of snprintf that is designed to be async-signal-safe. It supports a limited set of formats and does not initialize locale machinery.
* Set a flag and handle it in the main program: Instead of calling snprintf directly from a signal handler, set a flag inside the signal handler and have the main program check and handle the flag accordingly.
* Use write instead of printf: On POSIX-conforming systems, you can use write(STDOUT_FILENO, ...) instead of printf to output data from a signal handler. This approach may require additional formatting effort.

In summary, while snprintf is not inherently async-signal-safe, there are workarounds and alternatives available to ensure safe and reliable signal handling in C programs.

**Comment by Explorer09** [comment]
https://stackoverflow.com/questions/67839938/why-is-snprintf-not-considered-async-safe

**Comment by Explorer09** [comment]
Since the `CRT_handleSIGTERM` function simply exits after printing the diagnostic message, I wonder is the problem can be fixed by simply blocking other signals before calling `snprintf(3)`?

I mean, will the warning go away if we fix that way?

**Comment by BenBE** [comment]
This wont do …

**Comment by Explorer09** [comment]
Related to this bug. The `signal_safe_fprintf()` function in Settings.c is not quite "signal safe" as the function says, and should be fixed as well.

**Comment by DeltachangeOG** [comment]
Opened PR #1839 to address making handler async safe.

**Comment by DeltachangeOG** [comment]
I want to get feedback on what I’ve learned while digging into this.

At the moment, no part of the signal handling path is truly async-signal-safe. The existing handlers are intentionally written to terminate in-place (which is correct for preserving stack context, core dumps, and crash diagnostics), but they currently invoke libc, curses, and formatting code that can recurse, deadlock, or silently fail when a signal is delivered while execution is inside malloc or other libc internals.

I was able to reproduce this repeatedly: when a segmentation fault occurs while execution is inside malloc or other libc operations, the handler may produce no output at all (neither to the terminal nor stderr). This appears consistent with undefined behavior caused by re-entering non-reentrant code from a signal context.

Because of this, there isn’t a realistic way to make the current handlers fully signal-safe via small, localized changes alone. A strictly correct solution would require reworking signal handling more holistically.

That said, attempting to use the existing crash handler as a general-purpose reporting mechanism for all signals is not viable. The requirements for reliable crash reporting (terminate immediately, preserve stack context, minimal work in-handler) fundamentally conflict with the requirements for graceful shutdown and UI teardown.

There is, however, a conservative, incremental path forward that preserves existing invariants where they matter while improving reliability:

Proposed staged approach (minimal diff):

Clearly distinguish fatal signals (SIGSEGV, SIGILL, SIGFPE, SIGBUS, etc.) from non-fatal, user-initiated signals (SIGINT, SIGTERM, SIGQUIT).
Fatal signals must continue to terminate immediately in the signal handler to preserve stack context and core-dump behavior.
Non-fatal signals can be handled via a deferred shutdown path without impacting crash diagnostics.

Eliminate the most egregious non-signal-safe calls first (e.g. snprintf, strsignal) and replace them with constant writes selected via a simple switch on the signal number.
This reduces libc re-entry while preserving basic diagnostics.

Keep fatal signals terminating in place, exactly as they do today.

Defer broader restructuring (e.g. curses teardown or shared-state access) until there is agreement on the desired policy tradeoffs.

One additional clarification: deferring handling of non-fatal, user-initiated signals does not mask crashes, since fatal signals will still interrupt execution immediately and be handled in-place. Graceful shutdown paths cannot suppress or alter crash diagnostics.

Separately, for signals like SIGPIPE: since htop is an interactive UI tool, these typically represent non-fatal conditions (e.g. closed pipes or disconnected terminals). Treating SIGPIPE as ignorable or non-fatal aligns with common CLI/TUI behavior and avoids spurious crashes.

I’m willing to rework signal handling more extensively if that’s the preferred direction, but my immediate focus is on improving reliability by removing unsafe libc usage while preserving current crash semantics. If this approach sounds reasonable, I’ll rework the handler accordingly and submit a focused PR

**Comment by BenBE** [comment]
Sounds good.

**Comment by DeltachangeOG** [comment]
PR #1847 is the first step of the process, this cleanly separates signal handling and shutdown behavior, and makes the handler async-signal-safe. More extensive rewrites of the fatal signals will be forthcoming shortly.
